### PR TITLE
Use checkbox menu in crate selection.

### PR DESCRIPTION
**Pull request by poelzi**
Switch from normal QAction menu to a composition of QCheckBox and QActionWidget.
The checkbox shows in which crates the selection is in.
Changing the crates selection does not collapse the menu, which allows
much easier categorization of tracks without going through the menu from scratch.

### CODE REVIEW

**Comment by daschuer** [comment]
Thank you for this PR. Nice Idea.

It work well for a single track, unfortunately not well for multiple tracks. I think we need a three state check box or something, for the case where some of the selected tracks are in a crate and some not. 

Nit: Please move the pointer asterisk next to QWidget to meet our coding style.  

Be fore merge, we need your permission.
Please sign https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ and comment here when done.

@gramanas: how does this merge with your crate hierarchy branch?

**Comment by gramanas** [comment]
I can't check it atm, but this is based on master and not the new library redesign. Since it only changes widget files, I don't think it'll be hard to resolve the merge conflicts, but I can definitely see some.

**Comment by poelzi** [comment]
I implemented the tristate when multiple tracks are selected that do not share a crate fully.
Fixed your styling comment as well (I hope ;))

File: src/widget/wtracktableview.cpp
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: Ok, this works, but it took me quite a while to find out how. 
I think this requires some comments.
currentCrates should also be renamed ... mmm 
or can we move the size compare if condition to collectCrateTrackListOfTracks ?

File: src/widget/wtracktableview.cpp
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: contains() and a subsequent [] is a double hash lookup. Use find();

File: src/library/crate/cratestorage.cpp
Comment (daschuer) [reply src/library/crate/cratestorage.cpp]: this if condition is redundant, because operator[] has already such a code.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for the changes. 
It works nice now.

I have added some comments most of them are to improve the readability of the code.

File: src/widget/wtracktableview.cpp
Comment (poelzi) [reply src/widget/wtracktableview.cpp]: My thought was, that the API maybe more useful in the future. Giving a list of tracks and getting a hash with the crates and the containing tracks. Maybe I was overthinking it. I changed the api to just count the tracks in the crate.

File: src/widget/wtracktableview.cpp
Comment (poelzi) [reply src/widget/wtracktableview.cpp]: I tried different techniques for hours, but this was the only only I got working properly. I'm quite open if someone gets this running using more slick code, but as I think it's none critical code from performance aspects, I think it's OK.
I will write some more optimal SQL queries in a separate pull request which should be a much better optimization.

**Comment by poelzi** [comment]
Thanks for the review. I haven't user QT C++ before and haven't touched C++ for a very long time.
I hope I have addressed all points. Unfortunately I haven't found a way to avoid string properties in a slick way, but I don't think it's critical in any way.

File: src/library/crate/cratestorage.cpp
Comment (uklotzde) [inline src/library/crate/cratestorage.cpp]: Are you sure that the integer value in each entry is explicitly initialized with 0??? The fundamental type _int_ does not have a default constructor and QHash will leave it uninitialized when inserting a new key/value pair.

**Comment by uklotzde** [comment]
Very nice, I like this feature 👍 Now I can delete my custom branch that allowed me to delete selected tracks from a crate. But I omitted to add visual feedback, if those tracks are actually contained in crates. That's why I never created a PR.

**Comment by uklotzde** [comment]
Proposal: The name of the menu entry should be changed from "Add to Crate" to just "Crates"!

File: src/widget/wtracktableview.cpp
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: Unfortunately many queries are executed even if I don't open the _Crates_ menu entry. Each right click on selected tracks now has a noticeable delay when you have many crates. The list of crates with their state should only be populated when the corresponding sub-menu entry becomes visible.

File: src/library/crate/cratestorage.cpp
Comment (poelzi) [reply src/library/crate/cratestorage.cpp]: As far as I understood the QList documentation. It states that such types are initialized with 0.

File: src/widget/wtracktableview.cpp
Comment (poelzi) [reply src/widget/wtracktableview.cpp]: Ok, I will look into optimizing it and change the name to crates :)

File: src/widget/wtracktableview.cpp
Comment (daschuer) [reply src/widget/wtracktableview.cpp]: See https://github.com/poelzi/mixxx/pull/1

File: src/widget/wtracktableview.cpp
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: we have recently introduced "make_parented" to document the qt object tree ownership and to avoid the unused unique_ptr nature.

**Comment by uklotzde** [comment]
@poelzi Any plans to address the open comments?

I must admit that this is a really helpful feature that I merge into my private builds. I would like to see it in the next release(s). I usually have around 50+ crates and the performance penalties seem to be acceptable. We could even discuss to deliver the performance improvements in a follow-up release if (hopefully) we are able to return to a more frequent release cycle ;)

**Comment by poelzi** [comment]
I will hopefully get back addressing all the comments I have on my open pull requests and this one is annoying me as well ;) I had no working development machine for a while or was quite busy.

**Comment by poelzi** [comment]
I was able to optimize everything into one query with a subselect and reuse the CratesSummery code.
Got even slicker :)

**Comment by Be-ing** [comment]
@poelzi for the future, it's easier to review changes since the last review if new commits are added instead of rebasing.

File: src/library/crate/cratestorage.cpp
Comment (Be-ing) [inline src/library/crate/cratestorage.cpp]: nitpick: extra blank line here

File: src/library/crate/cratestorage.cpp
Comment (Be-ing) [inline src/library/crate/cratestorage.cpp]: Use standard C++11 `for` loop instead of old Qt `foreach` macro.

File: src/library/crate/cratestorage.cpp
Comment (Be-ing) [inline src/library/crate/cratestorage.cpp]: extra blank line

**Review by Be-ing — COMMENTED** [review]
just some small style nitpicks

**Comment by poelzi** [comment]
Be-ing: I think your nit picks where fixed by the optimization Uwe Klotz did.

**Comment by Be-ing** [comment]
Oh, you're correct. Sorry for the noise.

**Comment by uklotzde** [comment]
The idea of lazily populating the context sub-menus is also mentioned in this bug report:
https://bugs.launchpad.net/mixxx/+bug/1395022
Executing even more synchronous database queries to obtain results that might never be needed will impact the user experience even more.

Again, I like the functionality and would accept the lag in the first place. But we should think about how this can be fixed ASAP.

**Comment by poelzi** [comment]
@uklotzde i totally agree, but the current iteration does not make any more SQL calls  then before.

`sqlite> SELECT *, (    SELECT COUNT(*) FROM crate_tracks WHERE crates.id = crate_tracks.crate_id and crate_tracks.track_id in (29, 31, 28, 19, 15,135, 267) ) AS track_count, 0 as track_duration FROM crates ORDER BY name;
6|Haha1|0|1|0|0|3|0
4|blbl|0|1|0|0|0|0
3|blubb|0|1|0|0|0|0
5|huhu|0|1|0|0|3|0
10|huhuihaenuae|0|1|0|0|0|0
1|test1|0|1|1|0|2|0
2|test2|0|1|0|0|1|0
7|turaine|0|1|0|0|2|0
8|uaierutaors|0|1|0|0|0|0
9|uiaeuaie|0|1|0|0|1|0
-------- scanstats --------
Loop  1: SCAN TABLE crates USING INDEX sqlite_autoindex_crates_1
         nLoop=1        nRow=10       estRow=1048576  estRow/Loop=1.04858e+06
-------- subquery 1 -------
Loop  1: SEARCH TABLE crate_tracks USING COVERING INDEX sqlite_autoindex_crate_tracks_1 (crate_id=? AND track_id=?)
         nLoop=10       nRow=12       estRow=80       estRow/Loop=8       
---------------------------
Run Time: real 0.000 user 0.000000 sys 0.000000
sqlite> 
`
It's 2 loops instead of one but fully covered by indexes.

**Comment by uklotzde** [comment]
The lazy initialization of 2nd level actions affects multiple use cases and is a cross-cutting concern and could be solved in a separate task.

I vote for integrating this extremly useful feature before improving the performance of **all** context menus that require database access.

**Comment by Be-ing** [comment]
I test this and it works nicely. It would be cool to have this for Playlists too, but before that is implemented I think this needs to be lazy initialized or right clicking tracks will become even more terribly slow.

**Comment by uklotzde** [comment]
LGTM.

Some nitpicking about typos and coding style. But I don't think we need another round here. I would like to get the approval of 1 or 2 other reviewers before merging anything myself.

@daschuer How do we verify that the author has signed the contributor agreement?

**Comment by daschuer** [comment]
@poelzi is already a contributor.
Do you have a Launchpad account? Than I can add you to the Mixxx Contributors Group.

**Comment by poelzi** [comment]
@daschuer it's poelzi  on launchpad :)

**Review by Be-ing — CHANGES_REQUESTED** [review]

**Comment by uklotzde** [comment]
@poelzi Do you have time to address Be's comments and resolve the pending merge conflict, Daniel? I would like to have this in 2.1, because it has proven to be a valuable feature 👍

**Comment by Be-ing** [comment]
LGTM, thank you! @poelzi would you be interested in implementing the lazy initialization to alleviate https://bugs.launchpad.net/mixxx/+bug/1733200 for 2.1?

**Comment by poelzi** [comment]
@Be-ing kinda added the required infrastructure in another branch already. I added a widget (qlistview) that displays the crates of the selected track. I used a background thread to fetch the information in background and update the gui when done. If we decide to use one thread for sql queryies and gui support or one thread per component. https://ibb.co/dJsuf6

**Comment by Be-ing** [comment]
@uklotzde ready for merge?

File: src/widget/wtracktableview.cpp
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: This parameter should be passed as const-ref to avoid a deep copy of the list. Unfortunately the range-based for loops do not work well with Qt containers. Declaring the loop variable as const-ref is not enough. The container itself must be const! There is already a link in our [coding guidelines](https://www.mixxx.org/wiki/doku.php/coding_guidelines#range-based_for_loops).

We have this issue at many places and I have done this wrong until recently. Checking the code with [clazy](https://github.com/KDE/clazy) could reveal those hidden performance killers.

**Comment by uklotzde** [comment]
Almost ready. Just one minor issue that we should fix, now that we know better ;)

**Comment by Be-ing** [comment]
Thank you for this big usability improvements for crates! Let's continue this momentum to make Mixxx 2.2 the best music library management software. :)

File: src/widget/wtracktableview.cpp
Comment (daschuer) [reply src/widget/wtracktableview.cpp]: I will fix this after merge

Comment (Be-ing) [reply src/widget/wtracktableview.cpp]: I already did in #1427

Comment (daschuer) [reply src/widget/wtracktableview.cpp]: included in 187ec3d5fa231f2ee1259fe724ca50386a8ff9fc by @Be-ing

**Pull request merged**
